### PR TITLE
subquery uses document in not calc nodes

### PR DESCRIPTION
**Pull request by iurii-i-popov**
If a subquery uses document in not calc node do not apply optimization

### CODE REVIEW

**Comment by iurii-i-popov** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7682/

File: tests/js/server/aql/aql-optimizer-rule-late-document-materialization-arangosearch-2.js [hunk 1]
```diff
@@ -108,6 +108,11 @@ function lateDocumentMaterializationArangoSearch2RuleTestSuite () {
       let plan = AQL_EXPLAIN(query).plan;
       assertEqual(-1, plan.rules.indexOf(ruleName));
     },
+    testNotAppliedDueToUpdateDoc() {
```
Comment (Dronplane) [inline tests/js/server/aql/aql-optimizer-rule-late-document-materialization-arangosearch-2.js]: Suggest to check case with OLD/NEW in UPDATE/UPSERT queries

File: arangod/Aql/IndexNodeOptimizerRules.cpp [hunk 2]
```diff
@@ -74,26 +74,17 @@ namespace {
 
   void processCalculationNode(IndexNode const* indexNode, CalculationNode* calculationNode,
                               std::vector<latematerialized::NodeWithAttrs<latematerialized::AstAndFieldData>>& nodesToChange,
-                              bool NoSortNode, TRI_idx_iid_t& commonIndexId, bool& stickToSortNode, bool& stopSearch) {
+                              TRI_idx_iid_t& commonIndexId, bool& invalid) {
```
Comment (Dronplane) [inline arangod/Aql/IndexNodeOptimizerRules.cpp]: why not make method  to just return bool?

File: arangod/Aql/IndexNodeOptimizerRules.cpp [hunk 3]
```diff
@@ -157,47 +149,58 @@ void arangodb::aql::lateDocumentMaterializationRule(Optimizer* opt,
...
             break;
         }
-        // Currently only calculation and subquery nodes expected to use a loop variable.
-        // We successfully replaced all references to the loop variable.
-        // However if some other node types will begin to use the loop variable
-        // assertion below will be triggered and this rule should be updated.
-        if (!stopSearch && type != ExecutionNode::CALCULATION) {
+        // currently only calculation nodes expected to use a loop variable with attributes
+        // we successfully replace all references to the loop variable
+        if (!(stopSearch || invalid) && type != ExecutionNode::CALCULATION) {
```
Comment (Dronplane) [inline arangod/Aql/IndexNodeOptimizerRules.cpp]: It`s redundant to check stopSearch there - it is always false

Comment (Dronplane) [reply arangod/Aql/IndexNodeOptimizerRules.cpp]: My bad, REMOTE node could trigger it true.

**Comment by iurii-i-popov** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7696/

**Pull request merged**